### PR TITLE
DnS and DINO baselines

### PR DESCRIPTION
**Pull request by gkordo**
* Add code implementation of the two baselines
* Add a doc guide for each baseline
* Add a table in README.md with the baseline results on the training set

### CODE REVIEW

File: vsc/baseline/localization.py [hunk 1]
```diff
@@ -93,3 +94,60 @@ def score(self, candidate: CandidatePair, match: Match, box, similarity) -> floa
 class VCSLLocalizationCandidateScore(VCSLLocalization):
     def score(self, candidate: CandidatePair, match: Match, box, similarity) -> float:
         return candidate.score
+
+
+class VCSLLocalizationDnS(VCSLLocalizationMaxSim):
```
Comment (edpizzi) [inline vsc/baseline/localization.py:99]: If this is DnS specific, it might make sense to put it with DnS code -- maybe in dns_baseline? Or as a module within baseline/dns?

File: vsc/storage.py [hunk 2]
```diff
@@ -39,7 +39,7 @@ def same_value_ranges(values):
     yield value, start, len(values)
 
 
-def load_features(f, dataset: Optional[Dataset] = None):
+def load_features(f, dataset: Optional[Dataset] = None, as_dict: bool = False):
```
Comment (edpizzi) [inline vsc/storage.py:42]: Could we keep this method as is, and add a new function that translates `List[VideoFeature]` to a dict? I'd rather have a simple signature for this function.

File: vsc/baseline/dns/layers.py [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+        self.axis = axis
+        self.eps = eps
+
+    def __call__(self, module):
+        if hasattr(module, "weight"):
+            w = module.weight.data
+            module.weight.data = F.normalize(w, p=2, dim=self.axis, eps=self.eps)
+
+
+class Attention(nn.Module):
```
Comment (edpizzi) [inline vsc/baseline/dns/layers.py:21]: What do you think about using these models in torchscript format, so we don't need to include code for the layers? (Some pytorch code is difficult to use as a torchscript, but a lot of code works out of the box this way.)

Just an idea, but that's what we do for SSCD to avoid adding implementations here.

File: vsc/baseline/dns/students.py [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+            self.transformer = nn.TransformerEncoder(
+                encoder_layer, transformer_layers, nn.LayerNorm(dims)
+            )
+
+        if netvlad:
+            self.netvlad = NetVLAD(dims, netvlad_clusters, outdims=netvlad_outdims)
+
+        if pretrained:
+            self.load_state_dict(
+                torch.hub.load_state_dict_from_url(model_urls["dns_cg_student"])[
```
Comment (edpizzi) [inline vsc/baseline/dns/students.py:67]: Just a note -- candidates' code has to work without internet access. Our baselines don't necessarily have to conform to that, but it might be good to demonstrate that.

I guess the challenge here is that DnS requires a few models, and we don't have structured configs where we could specify several of them without a lot of plumbing.

I don't want to make this unnecessarily messy or difficult, so maybe we keep it as is.

**Review by edpizzi — COMMENTED** [review]
Just a few comments / questions. Thanks!

File: README.md [hunk 5]
```diff
@@ -37,6 +37,16 @@ $ ./matching_eval.py --predictions vsc_eval_data/matches.csv --ground_truth vsc_
 Matching track segment AP: 0.3650
 ```
 
+## Baseline results on training set
+
+| method                        | score norm         | Descriptor μAP | Matching μAP |
+|-------------------------------|--------------------|----------------|--------------|
+| [SSDC](docs/baseline.md)      | :x:                | 0.4754         | 0.3650       |
+| [SSDC](docs/baseline.md)      | :heavy_check_mark: | 0.6499         | 0.4692       |
+| [DINO](docs/baseline_dino.md) | :heavy_check_mark: | 0.4402         | 0.3393       |
```
[the remark this answers is not in the document]
Comment (gkordo) [reply README.md:46]: Sure. Just for clarity, the Descriptor μAP of DINO and DnS will be in italics.

File: docs/baseline_dino.md [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+method expects positive similarity in matching regions.
+
+The baseline script also runs a matching track evaluation, and a descriptor micro AP.
+The descriptor micro AP estimates the descriptor track score, but
+does not enforce the same limits as the `descriptor_eval.py` evaluation.
+
+The matching script provides several outputs in the `--output_path`
+directory, including precision-recall plots,
+score-normalized descriptor files, and both retrieved candidate pairs
+and localized matches in `.csv` formats.
```
[the remark this answers is not in the document]
Comment (gkordo) [reply docs/baseline_dino.md:92]: I agree. I will remove most chunks of duplicate text in the following commit. So I resolve this.

File: docs/installation.md [hunk 7]
```diff
@@ -9,6 +9,8 @@ We document the environments we've used for testing below:
 conda create --name vsc -c pytorch -c nvidia -c conda-forge pytorch \
   torchvision scikit-learn numpy pandas matplotlib faiss-gpu tqdm \
   pytorch-cuda=11.7
+conda activate vsc
+pip install einops
```
[the remark this answers is not in the document]
Comment (gkordo) [reply docs/installation.md:13]: Yes, it can be omitted.

File: vsc/baseline/localization.py (hunk 1 above)
Comment (gkordo) [reply vsc/baseline/localization.py:99]: This is indeed quite DnS specific method. I will move it to dns_baseline, since the baseline/dns will be removed if we go with the torchscript solution for DnS.

File: vsc/storage.py (hunk 2 above)
Comment (gkordo) [reply vsc/storage.py:42]: Sure, I have added a new function in the same .py to convert `List[VideoFeature]` to a dict, as suggested.

File: vsc/baseline/dns/layers.py (hunk 3 above)
Comment (gkordo) [reply vsc/baseline/dns/layers.py:21]: For DnS, that would work and simplify the code significantly.

But for DINO, it can not be easily converted to torchscript. So some code for this baseline is necessary.

File: vsc/baseline/dns/students.py (hunk 4 above)
Comment (gkordo) [reply vsc/baseline/dns/students.py:67]: This is resolved for DnS if we go with the torchscripts solution. 
But it still holds for DINO. So if this is not a big issue, I suggest leaving it as is.

File: vsc/baseline/dns/layers.py (hunk 3 above)
Comment (edpizzi) [reply vsc/baseline/dns/layers.py:21]: What is the failure with DINO? I believe I've converted DINO to torchscript in the past for the SSCD paper, but I no longer remember if I had to do anything tricky to accomplish this. :)

File: README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (edpizzi) [reply README.md:46]: Sounds good. I think that change is not applied yet.

File: vsc/storage.py (hunk 2 above)
Comment (edpizzi) [reply vsc/storage.py:42]: Great. I believe that change isn't reflected in the PR yet, but perhaps I've missed it

File: README.md [hunk 8]
```diff
@@ -37,6 +37,16 @@ $ ./matching_eval.py --predictions vsc_eval_data/matches.csv --ground_truth vsc_
 Matching track segment AP: 0.3650
 ```
 
+## Baseline results on training set
+
+| method                        | score norm         | Descriptor μAP | Matching μAP |
+|-------------------------------|--------------------|----------------|--------------|
+| [SSDC](docs/baseline.md)      | :x:                | 0.4754         | 0.3650       |
```
Comment (edpizzi) [inline README.md:44]: I missed this previously, but SSDC -> SSCD :)

File: vsc/baseline/inference.py [hunk 9]
```diff
@@ -30,6 +30,8 @@ class InferenceTransforms(enum.Enum):
     RESIZE_288 = enum.auto()
     # Resize the short edge to 320, then take the center crop
     RESIZE_320_CENTER = enum.auto()
+    # Aspect-ratio preserving resize to 224x224
```
Comment (edpizzi) [inline vsc/baseline/inference.py:33]: My read is that this doesn't preserve the aspect ratio, and compresses the long side to form a square. That's a good preprocessing choice for DINO, but the comment shouldn't say "aspect-ratio preserving". We can just say "Resize to 224x224".

File: README.md [hunk 10]
```diff
@@ -37,6 +37,16 @@ $ ./matching_eval.py --predictions vsc_eval_data/matches.csv --ground_truth vsc_
...
 ```
 
+## Baseline results on training set
+
+| method                        | score norm         | Descriptor μAP | Matching μAP |
+|-------------------------------|--------------------|----------------|--------------|
+| [SSCD](docs/baseline.md)      | :x:                | 0.4754         | 0.3650       |
+| [SSCD](docs/baseline.md)      | :heavy_check_mark: | 0.6499         | 0.4692       |
+| [DINO](docs/baseline_dino.md) | :heavy_check_mark: | _0.4402_       | 0.3393       |
+| [DnS](docs/baseline_dns.md)   | :heavy_check_mark: | _0.4129_       | 0.3211       |
```
Comment (edpizzi) [inline README.md:47]: Let's add a line below, something like "Numbers in _italics_ do not conform to challenge rules".

**Review by edpizzi — APPROVED** [review]
This looks great! Some minor documentation/comment suggestions. Very sorry for the delay.

Marking as approved. Feel free to land this using "squash and merge" with those changes, or I can land it.

File: vsc/baseline/inference.py (hunk 9 above)
Comment (gkordo) [reply vsc/baseline/inference.py]: Great catch. Fixed

File: README.md (hunk 10 above)
Comment (gkordo) [reply README.md:47]: Note added
